### step 1 : load and explore the dataset

In [68]:
import pandas as pd

In [70]:
movies = pd.read_csv("tmdb_5000_movies.csv/tmdb_5000_movies.csv")             ## importing the csv file  
credits = pd.read_csv("tmdb_5000_credits.csv/tmdb_5000_credits.csv")

In [72]:
movies.head()                         ## it shows the first 5 rows of the dataset movies

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [74]:
credits.head()               ## it shows the first 5 rows of the dataset credits

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

### step 2 : data cleaning and feature selection

In [77]:
movies = movies.merge(credits, on = "title")           ## here we merged the movies and credits into single dataset

In [79]:
movies.shape                        ## shape shows how many rows and columns does this dataset contains

(4809, 23)

In [81]:
movies.head()                       ## it shows the first 5 rows of the dataset 

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [83]:
movies = movies[["movie_id", "title", "overview", "genres", "keywords", "cast", "crew"]]   ## keeping the required columns 

In [85]:
## genres : Genres mean the category or type of the movie based on its theme, style, or story
## keywords : Keywords are important words or tags that describe the main themes, topics, or concepts of a movie
## overview : Overview is a short description or summary of what the movie is about — like the plot
## crew : People who work behind the scenes to make the movie, but don’t act in it(director, producer, writer , composer etc...)

In [87]:
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [89]:
movies.rename(columns = {"id" : "movie_id"}, inplace = True)    ## renaming the id with movie_id(if id is present instead of movie_id)

In [91]:
movies.isnull().sum()         ## checking for null values                 

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [93]:
movies.dropna(inplace = True)              ## droping the null values from the data

In [95]:
movies.reset_index(drop = True, inplace = True)        ## reindexing after removing the null value(reindex happened because after removing null values index suffled)

### step 3 : Data transformation  && step 4 : create tags for each movie

In [98]:
import ast

# Safer functions to handle bad data
def safe_convert(obj):
    try:
        parsed = ast.literal_eval(obj)
        return [i['name'] for i in parsed]
    except:
        return []

def safe_convert_cast(obj):
    try:
        parsed = ast.literal_eval(obj)
        return [i['name'] for i in parsed[:3]]  # take top 3 cast
    except:
        return []

def safe_fetch_director(obj):
    try:
        parsed = ast.literal_eval(obj)
        for i in parsed:
            if i['job'] == 'Director':
                return [i['name']]  # keep as list
        return []
    except:
        return []

# Apply all conversions
movies['genres'] = movies['genres'].apply(safe_convert)
movies['keywords'] = movies['keywords'].apply(safe_convert)
movies['cast'] = movies['cast'].apply(safe_convert_cast)
movies['crew'] = movies['crew'].apply(safe_fetch_director)
movies['overview'] = movies['overview'].apply(lambda x: x.split() if isinstance(x, str) else [])

# Now create tags properly by concatenating lists
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

# Test output
movies[['title', 'tags']].head()

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

### Step 4: Cleaning and Normalizing the Tags.

In [105]:
# Step 4: Clean and normalize the 'tags' column

# Convert list of words into a single string and lowercase everything
movies['tags'] = movies['tags'].apply(lambda x: " ".join(x))  # join the list
movies['tags'] = movies['tags'].apply(lambda x: x.lower())    # convert to lowercase

In [107]:
movies[['title', 'tags']].head()

title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

### Step 5: Text Vectorization using CountVectorizer

In [110]:
from sklearn.feature_extraction.text import CountVectorizer

# Initialize vectorizer
cv = CountVectorizer(max_features=5000, stop_words='english')

# Transform tags column into vectors
vectors = cv.fit_transform(movies['tags']).toarray()

In [112]:
vectors.shape

(4806, 5000)

### Step 6: Compute Similarity Using Cosine Similarity

In [115]:
from sklearn.metrics.pairwise import cosine_similarity

# Create similarity matrix
similarity = cosine_similarity(vectors)

In [117]:
similarity.shape

(4806, 4806)

###  Step 7: Build the Recommendation Function 🎬

In [120]:
def recommend(movie):
    # Get index of the movie from title
    movie_index = movies[movies['title'] == movie].index[0]
    
    # Get list of (index, similarity_score) for all movies
    distances = list(enumerate(similarity[movie_index]))
    
    # Sort based on similarity score in descending order
    movies_list = sorted(distances, reverse=True, key=lambda x: x[1])[1:6]
    
    # Print recommended movie titles
    for i in movies_list:
        print(movies.iloc[i[0]].title)

In [126]:
##checking the recommendation

In [122]:
recommend('Avatar')

Aliens
Moonraker
Alien
Alien³
Silent Running


In [124]:
movies['title'].head(20)

0                                          Avatar
1        Pirates of the Caribbean: At World's End
2                                         Spectre
3                           The Dark Knight Rises
4                                     John Carter
5                                    Spider-Man 3
6                                         Tangled
7                         Avengers: Age of Ultron
8          Harry Potter and the Half-Blood Prince
9              Batman v Superman: Dawn of Justice
10                               Superman Returns
11                              Quantum of Solace
12     Pirates of the Caribbean: Dead Man's Chest
13                                The Lone Ranger
14                                   Man of Steel
15       The Chronicles of Narnia: Prince Caspian
16                                   The Avengers
17    Pirates of the Caribbean: On Stranger Tides
18                                 Men in Black 3
19      The Hobbit: The Battle of the Five Armies


In [128]:
recommend("Spectre")

Quantum of Solace
Never Say Never Again
From Russia with Love
Skyfall
Die Another Day


In [134]:
recommend("The Dark Knight Rises")

The Dark Knight
Batman Begins
Batman
Batman
Batman Returns


## 🎬 Project Title: Movie Recommendation System
## 📌 Brief Description:
This project recommends movies to users based on a selected movie title. It uses content-based filtering and compares movies using text-based features like genres, keywords, cast, crew, and overview.

We extract and process movie metadata, convert it into numerical form using NLP, calculate similarity scores between movies, and recommend the top 5 similar ones.
## Steps You Completed:
### 1. Data Loading & Merging
Loaded two datasets: movies.csv and credits.csv

Merged them on the title column

### 2. Preprocessing
Selected relevant columns: movie_id, title, overview, genres, keywords, cast, and crew

Removed null values and duplicates

### 3. Data Transformation
Parsed JSON-like strings (genres, cast, etc.) into Python lists using ast.literal_eval

Extracted names (e.g., genre names, top 3 cast members, director)

Created a new column tags combining all key text info

### 4. Text Cleaning
Converted tags to lowercase

Joined word lists into strings for each movie

### 5. Vectorization (NLP)
Used CountVectorizer to convert text into vectors (top 5000 words)

Removed English stop words

Converted text into numerical matrix

### 6. Similarity Score Calculation
Used cosine similarity to compare every movie with every other movie

Created a similarity matrix of shape (4806, 4806)

### 7. Recommendation Function
Created a recommend(movie) function that:

Finds the movie

Finds its most similar movies

Prints the top 5 similar movie titles
## How It Works:
When a user inputs a movie name like "Avatar", the system:

Finds its vector in the dataset

Calculates similarity with all other movie vectors

Sorts by similarity score

Recommends the top 5 most similar movies (excluding the movie itself)